core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -2110,4 +2125,192 @@ public void testCurrentTranslogIDisCommitted() throws IOException {
+            CopyOnWriteArrayList<Throwable> failures = new CopyOnWriteArrayList<>();
+
+            // These threads add and block until the refresh makes the change visible and then do a non-realtime get.
+            Thread[] threads = new Thread[threadCount];
+            for (int i = 0; i < threadCount; i++) {
+                final String id = String.format(Locale.ROOT, "%04d", i);
+                final Term uid = newUid(id);
+                threads[i] = new Thread(() -> {
+                    int iteration = 0;
+                    while (run.get()) {
+                        try {
+                            iteration++;
+                            Document document = testDocument();
+                            String value = id + "i" + iteration; 
+                            document.add(new TextField("test", value, Field.Store.YES));
+                            ParsedDocument doc = testParsedDocument(id, id, "test", null, -1, -1, document, B_1, null);
+                            Engine.Index index = new Engine.Index(uid, doc);
+                            boolean created = engine.index(index);
+                            assertEquals(iteration, index.version());
+                            assertEquals(iteration == 1, created);
+
+                            DummyRefreshListener listener = new DummyRefreshListener(index.getTranslogLocation());
+                            engine.addRefreshListener(listener);
+                            assertBusy(() -> assertNotNull("listener never called", listener.forcedRefresh.get()));
+                            if (threadCount < defaultSettings.getMaxRefreshListeners()) {
+                                assertFalse(listener.forcedRefresh.get());
+                            }
+
+                            Engine.Get get = new Engine.Get(false, uid);
+                            try (Engine.GetResult getResult = engine.get(get)) {
+                                assertTrue("document not found", getResult.exists());
+                                assertEquals(iteration, getResult.version());
+                                SingleFieldsVisitor visitor = new SingleFieldsVisitor("test");
+                                getResult.docIdAndVersion().context.reader().document(getResult.docIdAndVersion().docId, visitor);
+                                assertEquals(Arrays.asList(value), visitor.fields().get("test"));
+                            }
+                        } catch (Throwable t) {
+                            failures.add(new RuntimeException("failure on the [" + iteration + "] iteration of thread [" + id + "]", t));
+                        }
+                    }
+                    logger.info("Finished [{}] iterations", iteration);
+                });
+                threads[i].start();
+            }
+            long end = System.nanoTime() + runTime;
+            while (System.nanoTime() < end && failures.isEmpty()) {
+                Thread.sleep(100);
```

Comment:
don't sleep here please use something like a latch or so that you can reset to resting and use operations